casper/src/main/resources/PoS.rhox

```diff
@@ -366,104 +507,97 @@ in {
-                                            @ListOps!("map", validatorsToWithdrawList, *computePay, *paymentsCh) |
-                                            contract computePay(@pk, resultCh) = {
-                                              resultCh!((pk, allBonds.get(pk) + rewards.getOrElse(pk, 0)))
-                                            } |
-                                            match validatorsToWithdrawList.toSet() {
-                                              validatorsToWithdrawSet => {
-                                                @ListOps!("filter", allBonds.toList(), *notWithdrawn, *newBondsListCh) |
-                                                @ListOps!("filter", withdrawers.toList(), *notWithdrawn, *newWithdrawersListCh) |
-                                                @ListOps!("filter", rewards.toList(), *notWithdrawn, *newRewardsListCh) |
-                                                for (@newWithdrawersList <- newWithdrawersListCh;
-                                                    @newBondsList       <- newBondsListCh;
-                                                    @newRewardsList     <- newRewardsListCh;
-                                                    @payments           <- paymentsCh) {
-                                                  returnCh!((
-                                                    payments,
-                                                    newBondsList.toMap(),
-                                                    newWithdrawersList.toMap(),
-                                                    // newRewardsList.toMap()
-                                                  ))
-                                                } |
-                                                // Checks that the validator is not a withdrawer.
-                                                contract notWithdrawn(@(pk, _), resultCh) = {
-                                                  resultCh!(not validatorsToWithdrawSet.contains(pk))
-                                                }
-                                              }
-                                            }
+                                    @ListOps!("filter", state.get("withdrawers").toList(), *isQuarantineFinished, *quarantinedValidatorsCh) |
+                                    for (@quarantinedValidators <- quarantinedValidatorsCh) {
+                                      @ListOps!("fold", quarantinedValidators, state, *computeRemove, *returnCh) |
+                                      contract computeRemove(@(pk, (bonds, quantine)), @updatedState, resultCh) = {
+                                        new payReturnCh in {
+                                          payWithdrawer!((pk, bonds + updatedState.get("comittedRewards").getOrElse(pk, 0)), *payReturnCh) |
+                                          for (@(true, _) <- payReturnCh) {
```

Comment:
In case of transfer error this code will hang.

The next question is, if we handle the error what to do. Should we continue or rollback everything.
Because state is updated in steps, rollback is not easy. For rollback it would be better to prepare all state changes and update the state as a final step.

Transfer can fail if PoS has not enough REVs, so we can just log the error and continue. I'm not sure what else can we do without heavier changes.